### ISSUE TITLE
LSTM

### ISSUE THREAD

**Issue by GriMacez** [GriMacez/-#1]
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error
from tensorflow.keras.models import Sequential, save_model, load_model
from tensorflow.keras.layers import LSTM, Dense, Dropout
from tensorflow.keras.callbacks import EarlyStopping
import os
import datetime

# 设置中文显示
plt.rcParams["font.family"] = ["SimHei", "WenQuanYi Micro Hei", "Heiti TC"]
plt.rcParams['axes.unicode_minus'] = False  # 解决负号显示问题

class TemperaturePredictor:
    def __init__(self, data_path, sequence_length=24, prediction_steps=12):
        """
        初始化温度预测器
        :param data_path: 数据路径
        :param sequence_length: 输入序列长度（小时）
        :param prediction_steps: 预测步数（小时）
        """
        self.data_path = data_path
        self.sequence_length = sequence_length
        self.prediction_steps = prediction_steps
        self.scaler = MinMaxScaler(feature_range=(0, 1))
        self.model = None
        self.data = None
        self.X_train, self.X_test, self.y_train, self.y_test = None, None, None, None
        self.X_train_lstm, self.X_test_lstm = None, None

    def load_data(self):
        """加载并预处理数据"""
        # 读取数据
        if self.data_path.endswith('.csv'):
            self.data = pd.read_csv(self.data_path)
        elif self.data_path.endswith('.xlsx') or self.data_path.endswith('.xls'):
            self.data = pd.read_excel(self.data_path)
        else:
            raise ValueError("数据格式不支持，请提供CSV或Excel文件")

        required_columns = ['DateTime', 'T (degC)']
        for col in required_columns:
            if col not in self.data.columns:
                raise ValueError(f"数据中缺少必要的列: {col}")

        # 转换日期时间格式
        self.data['DateTime'] = pd.to_datetime(self.data['DateTime'])

        # 按时间排序
        self.data.sort_values('DateTime', inplace=True)

        # 处理缺失值
        for col in self.data.select_dtypes(include=[np.number]).columns:
            # 先用向前填充
            self.data[col].fillna(method='ffill', inplace=True)
            # 再用向后填充处理开头的缺失值
            self.data[col].fillna(method='bfill', inplace=True)

        # 提取特征和目标变量
        features = ['T (degC)']
        if 'p (mbar)' in self.data.columns:
            features.append('p (mbar)')
        if 'rh (%)' in self.data.columns:
            features.append('rh (%)')

        self.feature_data = self.data[features].values

        return self.data

    def create_sequences(self, data, target_col_index=0):
        """
        创建序列数据用于LSTM
        :param data: 输入数据
        :param target_col_index: 目标列索引
        :return: 序列特征和目标
        """
        X, y = [], []
        for i in range(len(data) - self.sequence_length - self.prediction_steps + 1):
            # 输入序列
            X.append(data[i:i + self.sequence_length])

            # 目标序列（未来温度）
            y.append(data[i + self.sequence_length:i + self.sequence_length + self.prediction_steps, target_col_index])

        return np.array(X), np.array(y)

    def prepare_data(self):
        """准备LSTM模型的数据"""
        # 标准化数据
        scaled_data = self.scaler.fit_transform(self.feature_data)

        # 创建序列
        X, y = self.create_sequences(scaled_data)

        # 划分训练集和测试集
        self.X_train, self.X_test, self.y_train, self.y_test = train_test_split(
            X, y, test_size=0.2, shuffle=False
        )

        # 重塑数据以适应LSTM [样本数, 时间步, 特征数]
        self.X_train_lstm = self.X_train.reshape(self.X_train.shape[0], self.X_train.shape[1], self.X_train.shape[2])
        self.X_test_lstm = self.X_test.reshape(self.X_test.shape[0], self.X_test.shape[1], self.X_test.shape[2])

        print(f"训练数据形状: {self.X_train_lstm.shape}, {self.y_train.shape}")
        print(f"测试数据形状: {self.X_test_lstm.shape}, {self.y_test.shape}")

        return self.X_train_lstm, self.X_test_lstm, self.y_train, self.y_test

    def build_model(self):
        """构建LSTM模型"""
        model = Sequential()
        model.add(LSTM(128, return_sequences=True, input_shape=(self.sequence_length, self.X_train_lstm.shape[2])))
        model.add(Dropout(0.2))
        model.add(LSTM(64, return_sequences=False))
        model.add(Dropout(0.2))
        model.add(Dense(32))
        model.add(Dense(self.prediction_steps))  # 输出未来prediction_steps小时的温度

        model.compile(optimizer='adam', loss='mse', metrics=['mae'])
        self.model = model
        return model

    def train_model(self, epochs=50, batch_size=32, patience=10):
        """训练LSTM模型"""
        if self.model is None:
            self.build_model()

        # 设置回调函数
        callbacks = [
            EarlyStopping(monitor='val_loss', patience=patience, restore_best_weights=True)
        ]

        # 训练模型
        history = self.model.fit(
            self.X_train_lstm, self.y_train,
            epochs=epochs,
            batch_size=batch_size,
            validation_split=0.1,
            callbacks=callbacks,
            verbose=1
        )

        # 绘制训练历史
        self.plot_training_history(history)

        return history

    def plot_training_history(self, history):
        """绘制训练历史"""
        plt.figure(figsize=(12, 6))

        # 绘制损失
        plt.subplot(1, 2, 1)
        plt.plot(history.history['loss'], label='训练损失')
        plt.plot(history.history['val_loss'], label='验证损失')
        plt.title('模型损失')
        plt.xlabel('Epoch')
        plt.ylabel('损失')
        plt.legend()

        # 绘制MAE
        plt.subplot(1, 2, 2)
        plt.plot(history.history['mae'], label='训练MAE')
        plt.plot(history.history['val_mae'], label='验证MAE')
        plt.title('模型MAE')
        plt.xlabel('Epoch')
        plt.ylabel('MAE')
        plt.legend()

        plt.tight_layout()
        plt.savefig('training_history.png')
        plt.close()

    def evaluate_model(self):
        """评估模型性能"""
        if self.model is None:
            raise ValueError("模型尚未训练")

        # 在测试集上进行预测
        y_pred = self.model.predict(self.X_test_lstm)

        # 反标准化预测值和实际值
        # 为了反标准化，我们需要创建一个与原始数据形状相同的数组
        y_test_reshaped = np.zeros((len(self.y_test), self.feature_data.shape[1]))
        y_pred_reshaped = np.zeros((len(y_pred), self.feature_data.shape[1]))

        # 只填充温度列（第一列）
        y_test_reshaped[:, 0] = self.y_test[:, 0]
        y_pred_reshaped[:, 0] = y_pred[:, 0]

        # 反标准化
        y_test_inv = self.scaler.inverse_transform(y_test_reshaped)[:, 0]
        y_pred_inv = self.scaler.inverse_transform(y_pred_reshaped)[:, 0]

        # 计算评估指标
        rmse = np.sqrt(mean_squared_error(y_test_inv, y_pred_inv))
        mae = mean_absolute_error(y_test_inv, y_pred_inv)

        print(f"测试集RMSE: {rmse:.4f}")
        print(f"测试集MAE: {mae:.4f}")

        # 绘制预测结果
        self.plot_predictions(y_test_inv, y_pred_inv)

        return rmse, mae, y_test_inv, y_pred_inv

    def plot_predictions(self, y_test, y_pred):
        plt.figure(figsize=(15, 8))

        # 获取测试集对应的真实时间索引
        # 假设原始数据是 self.data，测试集起始索引可通过 train_test_split 的逻辑推导
        # 这里简化演示：假设测试集是原始数据的后 20%，取对应时间
        test_time_indices = self.data.index[-len(y_test):]
        real_dates = self.data['DateTime'].loc[test_time_indices]

        # 绘制前100个样本的预测结果（用真实时间做横轴）
        plt.plot(real_dates[:100], y_test[:100], label='实际温度')
        plt.plot(real_dates[:100], y_pred[:100], label='预测温度', alpha=0.7)

        plt.title('温度预测结果')
        plt.xlabel('真实时间')
        plt.ylabel('温度 (°C)')
        plt.xticks(rotation=45)  # 旋转日期，避免重叠
        plt.legend()
        plt.grid(True)

        plt.tight_layout()
        plt.savefig('temperature_predictions.png')
        plt.close()
    def predict_future(self, steps=12):
        """预测未来温度"""
        if self.model is None:
            raise ValueError("模型尚未训练")

        # 使用最后sequence_length个时间步的数据进行预测
        last_sequence = self.feature_data[-self.sequence_length:]
        scaled_sequence = self.scaler.transform(last_sequence)

        # 重塑数据以适应模型
        scaled_sequence = scaled_sequence.reshape(1, self.sequence_length, self.feature_data.shape[1])

        # 预测未来steps小时的温度
        future_temps_scaled = self.model.predict(scaled_sequence)

        # 反标准化预测结果
        future_temps_reshaped = np.zeros((len(future_temps_scaled[0]), self.feature_data.shape[1]))
        future_temps_reshaped[:, 0] = future_temps_scaled[0]
        future_temps = self.scaler.inverse_transform(future_temps_reshaped)[:, 0]

        # 生成未来时间标签
        last_date = self.data['DateTime'].iloc[-1]
        future_dates = [last_date + datetime.timedelta(hours=i + 1) for i in range(steps)]

        # 绘制预测结果
        plt.figure(figsize=(12, 6))
        plt.plot(self.data['DateTime'].iloc[-48:], self.data['T (degC)'].iloc[-48:], label='历史温度')
        plt.plot(future_dates, future_temps, 'r--', label='预测温度')

        plt.title(f'未来 {steps} 小时温度预测')
        plt.xlabel('日期时间')
        plt.ylabel('温度 (°C)')
        plt.legend()
        plt.grid(True)
        plt.xticks(rotation=45)

        plt.tight_layout()
        plt.savefig('future_temperature_prediction.png')
        plt.close()

        return future_dates, future_temps

# 主函数
if __name__ == "__main__":
    # 请替换为实际数据路径
    data_path = "data.csv"

    # 创建温度预测器实例
    predictor = TemperaturePredictor(data_path, sequence_length=24, prediction_steps=12)

    # 加载数据
    print("正在加载数据...")
    data = predictor.load_data()
    print(f"数据加载完成，共 {len(data)} 条记录")

    # 准备数据
    print("正在准备数据...")
    predictor.prepare_data()

    # 训练模型
    print("正在训练模型...")
    predictor.train_model(epochs=50, batch_size=32, patience=10)

    # 评估模型
    print("正在评估模型...")
    rmse, mae, y_test, y_pred = predictor.evaluate_model()

    # 预测未来12小时温度
    print("正在预测未来温度...")
    future_dates, future_temps = predictor.predict_future(steps=12)

    # 打印未来温度预测结果
    print("\n未来12小时温度预测结果:")
    for date, temp in zip(future_dates, future_temps):
        print(f"{date.strftime('%Y-%m-%d %H:%M')}: {temp:.2f}°C")
